### PR TITLE
reuse context for controller-manager base-context

### LINKED ISSUE #2701: controller: pass `ctx` parameter from top level to sub functions, instead of creating one new or using default context
**What would you like to be added**:
 pass `ctx` parameter from top level to sub functions, instead of creating one new or using default context

**Why is this needed**:
1. When create controllerManager, `Options.BaseContext` is empty:
https://github.com/karmada-io/karmada/blob/026781f207005c07b5f4cb0e76cf79b055bf53c0/cmd/controller-manager/app/controllermanager.go#L117

https://github.com/karmada-io/karmada/blob/026781f207005c07b5f4cb0e76cf79b055bf53c0/vendor/sigs.k8s.io/controller-runtime/pkg/manager/manager.go#L245

2. When create `Runnables`, `Options.BaseContext` empty, so `runnableGroup.ctx` will be `defaultBaseContext`:
https://github.com/karmada-io/karmada/blob/026781f207005c07b5f4cb0e76cf79b055bf53c0/vendor/sigs.k8s.io/controller-runtime/pkg/manager/manager.go#L616

https://github.com/karmada-io/karmada/blob/026781f207005c07b5f4cb0e76cf79b055bf53c0/vendor/sigs.k8s.io/controller-runtime/pkg/manager/runnable_group.go#L103

3. When run resource `reconcile` function, they use the `defaultBaseContext` :
https://github.com/karmada-io/karmada/blob/026781f207005c07b5f4cb0e76cf79b055bf53c0/vendor/sigs.k8s.io/controller-runtime/pkg/manager/runnable_group.go#L219

4. this `ctx` will be passed to resource `Reconcile` function:
https://github.com/karmada-io/karmada/blob/026781f207005c07b5f4cb0e76cf79b055bf53c0/pkg/controllers/namespace/namespace_sync_controller.go#L49

This will cause: `cancel signal` can't control some sub functions

### PR DESCRIPTION
**Pull request by yanggangtony**
**What type of PR is this?**
/kind cleanup
/kind feature

**What this PR does / why we need it**:
Reuse the context for Reconcile logic..

**Which issue(s) this PR fixes**:
Fixes #2701

### CODE REVIEW

File: cmd/controller-manager/app/controllermanager.go [hunk 1]
```diff
@@ -129,6 +132,7 @@ func Run(ctx context.Context, opts *options.Options) error {
 		LivenessEndpointName:       "/healthz",
 		MetricsBindAddress:         opts.MetricsBindAddress,
 		MapperProvider:             restmapper.MapperProvider,
+		BaseContext:                reUseContext,
```
Comment (jwcesign) [inline cmd/controller-manager/app/controllermanager.go:135]: Wouldn't it be better to write it this way?
```go
BaseContext: func() context.Context { return ctx }
```

Comment (yanggangtony) [reply cmd/controller-manager/app/controllermanager.go:135]: yep , that's great.
thanks.

**Comment by jwcesign** [comment]
cc @RainbowMango

**Comment by RainbowMango** [comment]
LGTM
This PR will close #2701, is there any other place we need to do the same？ @jwcesign

**Comment by jwcesign** [comment]
Hi, @yanggangtony , can you check all the `Reconcile` functions in `pkg/controllers`, some functions do not use the `ctx` parameter, such as `pkg/cluster/cluster_controller`:
```go
func (c *Controller) Reconcile(ctx context.Context, req controllerruntime.Request) (controllerruntime.Result, error) {
	klog.V(4).Infof("Reconciling cluster %s", req.NamespacedName.Name)

	cluster := &clusterv1alpha1.Cluster{}
        ############################### do not create one new ctx here ################
	if err := c.Client.Get(context.TODO(), req.NamespacedName, cluster); err != nil {
		// The resource may no longer exist, in which case we stop processing.
		if apierrors.IsNotFound(err) {
			return controllerruntime.Result{}, nil
		}

		return controllerruntime.Result{Requeue: true}, err
	}

	if !cluster.DeletionTimestamp.IsZero() {
		return c.removeCluster(ctx, cluster)
	}

	return c.syncCluster(ctx, cluster)
}
```

**Comment by yanggangtony** [comment]
> Hi, @yanggangtony , can you check all the `Reconcile` functions in `pkg/controllers`, some functions do not use the `ctx` parameter, such as `pkg/cluster/cluster_controller`:

@jwcesign 
you means confirm and update all Reconcile() method in pkg/controllers to use the ctx parameter,?

**Comment by jwcesign** [comment]
> you means confirm and update all Reconcile() method in pkg/controllers to use the ctx parameter?

Yes, thanks a lot!

**Comment by yanggangtony** [comment]
> > you means confirm and update all Reconcile() method in pkg/controllers to use the ctx parameter?
> 
> Yes, thanks a lot!

OK ,I will do it now.

**Comment by yanggangtony** [comment]
@jwcesign 
PTAL

**Comment by jwcesign** [comment]
Nice work!
/lgtm

**Comment by RainbowMango** [comment]
We should do the same for `karmada-agent`:
```diff
$ git diff
diff --git a/cmd/agent/app/agent.go b/cmd/agent/app/agent.go
index b6b7a016..11ea5357 100644
--- a/cmd/agent/app/agent.go
+++ b/cmd/agent/app/agent.go
@@ -190,6 +190,9 @@ func run(ctx context.Context, karmadaConfig karmadactl.KarmadaConfig, opts *opti
                LivenessEndpointName:       "/healthz",
                MetricsBindAddress:         opts.MetricsBindAddress,
                MapperProvider:             restmapper.MapperProvider,
+               BaseContext: func() context.Context {
+                       return ctx
+               },
                Controller: v1alpha1.ControllerConfigurationSpec{
                        GroupKindConcurrency: map[string]int{
                                workv1alpha1.SchemeGroupVersion.WithKind("Work").GroupKind().String():       opts.ConcurrentWorkSyncs,
```

**Comment by yanggangtony** [comment]
@RainbowMango 
Hi , thanks for review.
had updated karmada-agent's code

**Review by RainbowMango — APPROVED** [review]
/lgtm
/approve
Thanks for your quick response.

**Pull request merged**